source/httpfs/driver.go

```diff
@@ -37,16 +24,35 @@ type Driver struct {
 // this driver. This reduces the number of error handling branches in clients of
 // this package without compromising on correctness.
 func New(fs http.FileSystem, path string) source.Driver {
-	var d Driver
+	var d driver
 	if err := d.Init(fs, path); err != nil {
 		return &failedDriver{err}
```

Comment:
It is all for the convenience of the users of this package. We have roughly dozen of more services that are using migrate package to automatically apply up migrations on service start. Other services apply migrations on demand (by receiving a request via API). And none of them would prefer handing migration errors twice - once for creating the driver and then for performing the actual migration.

Initial implementation used to return error here. But the client code were always identical - create driver report error if necessary, perform migrations report error if necessary. Looking from the higher lever we do not really care why we can not run migrations (problem reading source, error accessing DB, malformed migration file, duplicate migrations). We would like to know if the operation succeed or not.

This pushed me to add delayed error handling. The nice thing about it is that we are not sacrificing consistency or correctness by doing that. If someone created a driver and never used it - there is not harm not reporting an error. It was actually used the error will be reported.

It could reinforced with the following thought experiment. Imagine we rewrote the `Init()` function to only store given `fs` and `path` variables internally but do not perform any directory listing. Then on first usage call like `Next()`, `First()`, `Prev()`... we would perform the directory listing, cache it and return the result. Subsequent calls uses data from the cache. In this scenario behaviour would be identical. No errors possible on `Init()` and dir listing error reported on first usage call. It is totally up to the source to decide when directory listing is actually performed. The only reason package uses deferred errors instead of lazy directory listing is to avoid concurrency issues on the first call (concurrent calls could race for data listing).

If I still have not managed to convince you to leave deferred error handling just let me know. I and will rip it away from this package without any further excuses :smile:.

Reply:
>Other services apply migrations on demand (by receiving a request via API). And none of them would prefer handing migration errors twice - once for creating the driver and then for performing the actual migration.
> ...
> Looking from the higher lever we do not really care why we can not run migrations (problem reading source, error accessing DB, malformed migration file, duplicate migrations). We would like to know if the operation succeed or not.

This seems specific to your usage of `migrate`. It's fine for you to hide the error at the service API level, but you probably want to log the details of the error.

In general, deferring the errors make it harder to debug when you actually encounter the error. e.g. You'd first have to realize that the error is coming from `failedDriver` and figure out why that's happening.

If you'd like to defer errors, please do that in your own `http.FileSystem` driver implementation. For an open source package that's used by different teams with different needs, it's better for the interfaces to be simple, clear, and easy to debug.

Reply:
:+1:  done.